pkg/service/admission/externalip_admission.go

```diff
@@ -0,0 +1,68 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	kadmission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+)
+
+const ExternalIPPluginName = "ExternalIPRanger"
+
+func init() {
+	kadmission.RegisterPlugin("ExternalIPRanger", func(client clientset.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewExternalIPRanger(nil), nil
+	})
+}
+
+type externalIPRanger struct {
+	*kadmission.Handler
+	network *net.IPNet
+}
+
+var _ kadmission.Interface = &externalIPRanger{}
+
+// NewConstraint creates a new SCC constraint admission plugin.
+func NewExternalIPRanger(network *net.IPNet) *externalIPRanger {
+	return &externalIPRanger{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		network: network,
+	}
+}
+
+// Admit determines if the service should be admitted based on the configured network CIDR.
+func (r *externalIPRanger) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != kapi.Resource("services") {
+		return nil
+	}
+
+	svc, ok := a.GetObject().(*kapi.Service)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	var errs field.ErrorList
+	switch {
+	// administrator disabled externalIPs
+	case len(svc.Spec.ExternalIPs) > 0 && r.network == nil:
+		errs = append(errs, field.Forbidden(field.NewPath("spec", "externalIPs"), "externalIPs have been disabled"))
+	// administrator has limited the range
+	case len(svc.Spec.ExternalIPs) > 0 && r.network != nil:
+		for i, s := range svc.Spec.ExternalIPs {
+			ip := net.ParseIP(s)
+			if ip == nil || !r.network.Contains(ip) {
+				errs = append(errs, field.Forbidden(field.NewPath("spec", "externalIPs").Index(i), fmt.Sprintf("externalIPs must be in the CIDR range %s", r.network)))
```

Comment:
We generally use forbidden when we disable a field for some security reason.

On Tue, Mar 8, 2016 at 8:02 AM, David Eads <notifications@github.com> wrote:

> In pkg/service/admission/externalip_admission.go
> <https://github.com/openshift/origin/pull/7810#discussion_r55361720>:
>
> > +	// if we can't convert then we don't handle this object so just return
> > +	if !ok {
> > +		return nil
> > +	}
> > +
> > +	var errs field.ErrorList
> > +	switch {
> > +	// administrator disabled externalIPs
> > +	case len(svc.Spec.ExternalIPs) > 0 && r.network == nil:
> > +		errs = append(errs, field.Forbidden(field.NewPath("spec", "externalIPs"), "externalIPs have been disabled"))
> > +	// administrator has limited the range
> > +	case len(svc.Spec.ExternalIPs) > 0 && r.network != nil:
> > +		for i, s := range svc.Spec.ExternalIPs {
> > +			ip := net.ParseIP(s)
> > +			if ip == nil || !r.network.Contains(ip) {
> > +				errs = append(errs, field.Forbidden(field.NewPath("spec", "externalIPs").Index(i), fmt.Sprintf("externalIPs must be in the CIDR range %s", r.network)))
>
> Seems more Invalid than Forbidden, but I don't feel strongly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7810/files#r55361720>.
>